sample_files/main.py

```diff
@@ -1,145 +1,80 @@
-        ("human", html_src_prompt),
-        ("human", "{urls}")
-    ]
-)
-
-# Setting up python_repl that runs valid python command
-python_repl = PythonREPL()
-repl_tool = Tool(
-    name = "python_repl",
-    description = '''A Python shell. Use this to execute python commands. 
-    Input should be a valid python command. Wait for code to execute completely ''',
-    func = python_repl.run,
-)
-
+load_dotenv()
 
-def save_and_process(output: str):
-    """Save output to a file and process it with repl_tool."""
-    # Save the sanitize_output to a file
-    if output is not None:
-        with open("output_scripts/selenium_script.py", "w+") as f:
-            f.write(output)
-    
-    # Process the output with the lambda function
-    return repl_tool(output) if output is not None else None
+# Fetch API key
+api_key = os.getenv('OPENAI_API_KEY')
 
- 
-# StrOutputParser() - converts response into parsable string
-chain = (
-        prompt |
-        llm | 
-        StrOutputParser() | 
-        sanitize_output |
-        save_and_process
-        )
+# Initialize OpenAI Chat model
+llm = ChatOpenAI(temperature=0.8, api_key=api_key)
 
+# Initialize Chrome WebDriver using WebDriver Manager
+def get_webdriver():
+    return webdriver.Chrome(ChromeDriverManager().install())
 
-def check_chain_interactions(chain_results,urls)-> bool:
-    """Check if elements from chain_results are present on each webpage in urls."""
+# Function to check interactions with Selenium WebDriver
+def check_chain_interactions(chain_results, urls):
     for url in urls:
-      try:
-          # Initialize Selenium WebDriver
-          driver = webdriver.Chrome()  
-          driver.get(url)  
-
-          # Check if the elements interacted with by the chain are present on the webpage
-          for result in chain_results:
-              element = driver.find_element(By.XPATH, f"//*[contains(text(), '{result}')]")
-              if not element:
-                  logger.error(f"Element {result} not found on the webpage")
-                  return False
-
-          driver.quit()
-          return True
-
-      except WebDriverException as e:
-          logger.error(f"Selenium WebDriver error: {e}")
-          return False
-      
-
-  
-def execute_chain(urls:list, max_attempts=5, attempt=1):
-    """Execute a chain process, retries if it fails, up to max_attempts."""
+        try:
+            driver = get_webdriver()
+            driver.get(url)
+            for result in chain_results:
+                element = driver.find_element(By.XPATH, f"//*[contains(text(), '{result}')]")
+                if not element:
+                    logger.error(f"Element {result} not found on the webpage {url}")
+                    driver.quit()
+                    return False
+            driver.quit()
+            return True
+        except WebDriverException as e:
+            logger.error(f"Selenium WebDriver error: {e}")
+            return False
+
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

In the `check_chain_interactions` function, the WebDriver is instantiated for each URL in the loop. This can be inefficient if there are many URLs to check. Consider moving the instantiation outside the loop and reusing the same driver instance for all URLs.

```diff
- def check_chain_interactions(chain_results, urls):
-     for url in urls:
-         try:
-             driver = get_webdriver()
-             driver.get(url)
-             ...
-             driver.quit()
-             return True
-         except WebDriverException as e:
-             logger.error(f"Selenium WebDriver error: {e}")
-             return False
+ def check_chain_interactions(chain_results, urls):
+     try:
+         driver = get_webdriver()
+         for url in urls:
+             driver.get(url)
+             ...
+         driver.quit()
+         return True
+     except WebDriverException as e:
+         logger.error(f"Selenium WebDriver error: {e}")
+         return False
```



<!-- This is an auto-generated comment by OSS CodeRabbit -->